### PR TITLE
Add XML caching functionality (based on bradsdavis and lincolnthree's work in this area)

### PR DESCRIPTION
**Pull request by jsight**
(will mark for review when I finish looking at XmlUtil)

### CODE REVIEW

File: rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XMLDocumentCache.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    private static final Map<String, CacheDocument> map = new HashMap<>();
+
+    /**
+     * This is used to pass data back to the caller regarding the cache search.
+     *
+     * If a parsing failure had been cached, {@link Result#isParseFailure()} will return true, and {@link Result#getDocument()} will return false.
+     *
+     * A cache miss will have {@link Result#isParseFailure()} set to false and {@link Result#getDocument()} will be null.
+     */
+    public static class Result
```
Comment (lincolnthree) [inline rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XMLDocumentCache.java]: Ah, good idea!

File: rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XMLDocumentCache.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    private static String getKey(XmlFileModel key)
+    {
+        return key.getFilePath();
+    }
+
+    private static class CacheDocument
+    {
+        private boolean parseFailure;
```
Comment (lincolnthree) [inline rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XMLDocumentCache.java]: Also a good idea!

File: rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XmlFileModel.java [hunk 3]
```diff
@@ -52,20 +55,35 @@
...
-                throw new WindupException("Exception reading document.", e);
+                FileModel fileModel = frame(asVertex(), FileModel.class);
+                try (InputStream is = fileModel.asInputStream())
+                {
+                    document = LocationAwareXmlReader.readXML(is);
+                    XMLDocumentCache.cache(this, document);
+                }
+                catch (Exception e)
+                {
+                    XMLDocumentCache.cacheParseFailure(this);
```
Comment (lincolnthree) [inline rules-xml/addon/src/main/java/org/jboss/windup/rules/apps/xml/model/XmlFileModel.java]: Sure.

File: utils/src/main/java/org/jboss/windup/util/xml/XmlUtil.java [hunk 4]
```diff
@@ -73,6 +82,9 @@ public static String nodeToString(Node node)
         return sw.toString();
     }
 
+    /**
+     * Returns the names and locations defined in this {@link Document}.
```
Comment (lincolnthree) [inline utils/src/main/java/org/jboss/windup/util/xml/XmlUtil.java]: Awesome.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-pr-builder/696/
Test PASSed.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-core-PR-dependents-check/68/
Test PASSed.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-pr-builder/697/
Test PASSed.

**Comment by windupgithubbot1** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://10.3.8.5//job/windup-core-PR-dependents-check/69/
Test PASSed.

**Pull request merged**